### ISSUE TITLE
EPIC: Push/Pull Mempool

### ISSUE THREAD

**Issue by evan-forbes** [celestiaorg/celestia-core#884]
This is a placeholder issue as we flesh out what the push pull mempool will look like.

In order to increase our block size, we will need to dramatically increase tendermint's p2p stack efficiency. There are many different things that we could do this, one proposed mechanism is to have the mempool gossip transactions and rollup block data like normal the first time a node sees that data, but then only gossip hashes should that tx or blob stay in the mempool. If a peer recieves a hash of a tx or blob that it has not seen before, then it can request the data for that hash from the peer who sent the hash. This functionality would be needed for #883.

**Comment by cmwaters** [comment]
Update: I have a working prototype of this on [`feature/cat`](https://github.com/celestiaorg/celestia-core/tree/feature/cat) which builds off `v1.5.0...`. The current specification can be found [here](https://github.com/celestiaorg/celestia-core/blob/feature/cat/mempool/cat/spec.md). I intend to start working on some performance benchmarks to compare this mempool with the priority mempool (v1) in `v1.5.0...`/`v0.10.0` (app).

**Comment by cmwaters** [comment]
Synced with @Bidon15 on the analysis gameplan. Since testground doesn't yet support metrics, I've modified the mempool reactors to dump all relevant metrics to a JSON file in the `data` directory every 10 seconds. I've verified that this is working with the e2e tests (simple 4 validator network run for ~12 minutes). Comparing the v1 (prioritized mempool) with v2 (content addressable transaction pool) will give us the following data:

v1 (validator02)
```
{
"StartedAt":"2022-12-15T13:30:23.70262411Z",
"EndedAt":"2022-12-15T13:42:53.771855021Z",
"FailedTxs":0,
"EvictedTxs":0,
"SuccessfulTxs":741,
"AlreadySeenTxs":1130,
"SentTransactionBytes":3428100,
"SentStateBytes":0,
"ReceivedTxBytes":3494090,
"ReceivedStateBytes":0
}
```

v2 (validator03)
```
{
"StartedAt":"2022-12-15T13:15:34.843873053Z",
"EndedAt":"2022-12-15T13:27:44.880283237Z",
"FailedTxs":3,
"EvictedTxs":0,
"SuccessfulTxs":704,
"AlreadySeenTxs":4,
"SentTransactionBytes":1163337,
"SentStateBytes":79548,
"ReceivedTxBytes":1076671,
"ReceivedStateBytes":79674
}
```

Here each tx is 1kb. With testground, we will initially be testing a 100 validator network with 1000 clients mostly using PFB/Blob txs. The test will span roughly 10 minutes. 

The main gain from the content addressable transaction pool that we expect are reductions in bandwidth consumption. While this in itself is appealing to validators, we should measure the impact such bandwidth reductions have on the following baselines:
- transaction throughput (This is roughly the bytes agreed upon per 100 validators per second)
- transaction latency (The mean time between transaction submission and finalization)
- block production variance (standard deviation as well as range in block times. This is often influenced by timeouts under load. We need to ensure that the network makes progress reliably even under load)
